src/ctap/mod.rs

```diff
@@ -165,19 +166,54 @@ pub struct AssertionState {
 pub enum StatefulCommand {
     Reset,
     GetAssertion(AssertionState),
-    EnumerateRps(Vec<String>),
+    EnumerateRps(usize),
     EnumerateCredentials(Vec<usize>),
 }
 
-pub fn check_command_permission(
-    stateful_command_permission: &mut TimedPermission,
-    now: ClockValue,
-) -> Result<(), Ctap2StatusCode> {
-    *stateful_command_permission = stateful_command_permission.check_expiration(now);
-    if stateful_command_permission.is_granted(now) {
-        Ok(())
-    } else {
-        Err(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED)
+pub struct StatefulPermission {
+    permission: TimedPermission,
+    command_type: Option<StatefulCommand>,
+}
+
+impl StatefulPermission {
+    // Resets are only possible in the first 10 seconds after booting.
+    // Therefore, initialization includes allowing Reset.
+    pub fn new_reset(now: ClockValue) -> StatefulPermission {
+        StatefulPermission {
+            permission: TimedPermission::granted(now, RESET_TIMEOUT_DURATION),
+            command_type: Some(StatefulCommand::Reset),
+        }
+    }
+
+    pub fn clear(&mut self) {
+        self.permission = TimedPermission::waiting();
+        self.command_type = None;
+    }
+
+    pub fn check_command_permission(&mut self, now: ClockValue) -> Result<(), Ctap2StatusCode> {
+        if self.permission.is_granted(now) {
+            Ok(())
+        } else {
+            self.clear();
+            Err(Ctap2StatusCode::CTAP2_ERR_NOT_ALLOWED)
+        }
+    }
+
+    pub fn get_command(&mut self) -> Result<&mut StatefulCommand, Ctap2StatusCode> {
```

Comment:
Return a mutable value is not the typical idiom you expect for a getter.
Why not having `get_command` return a immutable and use `set_command` to modify the value? If the issue is the timestamp that shouldn't be updated, you may also add `update_command` that calls `set_command` with the stored timestamp.

Reply:
Instead of one `update_command` function, I now have individual update functions to handle the correct states. This makes the API even more strict, as commands can't update their state outside more than necessary, e.g. to a different command. Updating now always restarts the whole command permissions.